File: typescipt/src/compiler/cSharp/builtInFunctions/lexyFunctionCall.ts


export class LexyFunctionCall extends FunctionCall {
   public LexyFunction ExpressionFunction

   public LexyFunctionCall(LexyFunction expressionFunction) : base(expressionFunction) {
     ExpressionFunction = expressionFunction;
   }

   public override customMethodSyntax(context: ICompileFunctionContext): MemberDeclarationSyntax {
     return null;
   }

   public override callExpressionSyntax(context: ICompileFunctionContext): ExpressionSyntax {
     return RunFunction(ExpressionFunction.FunctionName, ExpressionFunction.VariableName);
   }

   public static runFunction(functionName: string, variableName: string): InvocationExpressionSyntax {
     return SyntaxFactory.InvocationExpression(
         SyntaxFactory.MemberAccessExpression(
           SyntaxKind.SimpleMemberAccessExpression,
           SyntaxFactory.IdentifierName(ClassNames.FunctionClassName(functionName)),
           SyntaxFactory.IdentifierName(LexyCodeConstants.RunMethod)))
       .WithArgumentList(
         SyntaxFactory.ArgumentList(
           SyntaxFactory.SeparatedArray<ArgumentSyntax>(
             new SyntaxNodeOrToken[] {
               SyntaxFactory.Argument(SyntaxFactory.IdentifierName(variableName)),
               SyntaxFactory.Token(SyntaxKind.CommaToken),
               SyntaxFactory.Argument(SyntaxFactory.IdentifierName(LexyCodeConstants.ContextVariable))
             })));
   }
}

File: dotnet/Lexy/Lexy.Compiler/Compiler/CSharp/BuiltInFunctions/LexyFunctionCall.cs
using Lexy.Compiler.Language.Expressions.Functions;
using Microsoft.CodeAnalysis;
using Microsoft.CodeAnalysis.CSharp;
using Microsoft.CodeAnalysis.CSharp.Syntax;

namespace Lexy.Compiler.Compiler.CSharp.BuiltInFunctions;

internal class LexyFunctionCall : FunctionCall<LexyFunction>
{
    public override ExpressionSyntax CallExpressionSyntax(LexyFunction expression)
    {
        return RunFunction(expression.FunctionName, expression.VariableName);
    }

    public static InvocationExpressionSyntax RunFunction(string functionName, string variableName)
    {
        return SyntaxFactory.InvocationExpression(
                SyntaxFactory.MemberAccessExpression(
                    SyntaxKind.SimpleMemberAccessExpression,
                    SyntaxFactory.IdentifierName(ClassNames.FunctionClassName(functionName)),
                    SyntaxFactory.IdentifierName(LexyCodeConstants.RunMethod)))
            .WithArgumentList(
                SyntaxFactory.ArgumentList(
                    SyntaxFactory.SeparatedList<ArgumentSyntax>(
                        new SyntaxNodeOrToken[]
                        {
                            SyntaxFactory.Argument(SyntaxFactory.IdentifierName(variableName)),
                            SyntaxFactory.Token(SyntaxKind.CommaToken),
                            SyntaxFactory.Argument(SyntaxFactory.IdentifierName(LexyCodeConstants.ContextVariable))
                        })));
    }
}